src/hotspot/share/prims/jvmtiEventController.cpp

```diff
@@ -443,7 +457,13 @@ JvmtiEventControllerPrivate::recompute_env_enabled(JvmtiEnvBase* env) {
   }
 
   // will we really send these events to this env
-  env->env_event_enable()->_event_enabled.set_bits(now_enabled);
+  if (((was_enabled & OBJECT_FREE_BIT) != 0) ||
+      ((now_enabled & OBJECT_FREE_BIT)) != 0) {
+    // Set/reset the event enabled under the tagmap lock.
+    set_enabled_events_with_lock(env, now_enabled);
```

Comment:
You could tighten up the test to only handle specially when the state of the ObjectFree event is changing, i.e.

```if (((was_enabled ^ now_enabled) & OBJECT_FREE_BIT) != 0)```

Or you could not bother with the conditionalization at all, and just always call set_enabled_events_with_lock; I bet nobody would notice any performance difference.  That would eliminate "benign" races between unlocked bit setting here and bit testing in remove_dead_entries_locked.  Of course, the current implementation has such races on these bits all over the place; what's one race more or less among friends...

Or you could just leave it as you have it.  Your call.

Reply:
I like the idea of reseting the enabled bits under a lock unconditionally.  I'll do that.